lib/ravelin/password.rb

```diff
@@ -0,0 +1,27 @@
+require 'digest'
+
+module Ravelin
+  class Password < RavelinObject
+    attr_accessor :success, :failure_reason, :password_hashed
+    attr_required :success
+
+    # Alternative interface, because when the attr is called "password_hashed",
+    # the end user might think they need to hash the password themselves
+    def password=(passwd)
+      @password_hashed = Digest::SHA256.hexdigest(passwd)
+    end
+
+    def password_hashed=(passwd)
+      @password_hashed = passwd
+    end
+
+    FAILURE_REASONS = %w(BAD_PASSWORD UNKNOWN_USERNAME INTERNAL_ERROR RATE_LIMIT)
+
+    def validate
+      super
+      if !success && !FAILURE_REASONS.include?(failure_reason)
+        raise InvalidFailureReasonError.new("Failure reason value must be one of #{FAILURE_REASONS.join(', ')}")
```

Comment:
Rather than defining a new exception type here, I'd just raise `ArgumentError`. In general you only need to raise a specific exception type if there's a specific way to handle it, but here there isn't: Setting an invalid failure reason is a programming error which can't reasonably be handled, and would need a code fix.